src/rust/engine/fs/src/fs_tests.rs

```diff
@@ -0,0 +1,35 @@
+// Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
+// Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+use super::GitignoreStyleExcludes;
+use super::PathGlob;
+use super::PathGlobs;
+use crate::{GlobExpansionConjunction, StrictGlobMatching};
+
+#[test]
+fn path_globs_create_distinguishes_between_includes_and_excludes() {
+  let include_globs = ["foo.rs".to_string(), "bar.rs".to_string()];
+  let parsed_exclude_globs = ["ignore.rs".to_string(), "**/*.rs".to_string()];
+
+  let mut glob_inputs: Vec<String> = vec![];
+  glob_inputs.extend_from_slice(&include_globs);
+  glob_inputs.extend(parsed_exclude_globs.iter().map(|glob| format!("!{}", glob)));
+
+  let pg = PathGlobs::create(
+    glob_inputs.as_slice(),
+    StrictGlobMatching::Ignore,
+    GlobExpansionConjunction::AllMatch,
+  )
+  .expect("Path globs failed to be created");
+
+  assert_eq!(
+    pg.include,
+    PathGlob::spread_filespecs(&include_globs).expect("Include globs failed to expand")
+  );
+  assert_eq!(
+    pg.exclude.patterns,
+    GitignoreStyleExcludes::create(&parsed_exclude_globs)
+      .expect("Exclude globs failed to expand")
+      .patterns
+  );
```

Comment:
This was harder to write than I expected, mostly due to being green with Rust still. Thanks to @gshuflin for holding my hand through it!

Reply:
I'm super good at pairing on these things in the future as well if Greg happens to be working on something else at the time (or if you'd ever want a third participant)! Rust has the definition of a learning curve but I swear to you the simple beauty continues to overwhelm you with each step!

Reply:
Thanks for the offer! (And the Rust help over the weekend)

FTR, what I struggled the most with here to do the equivalent of `[*include_globs, *(f"!{glob}" for glob in exclude_globs)]`. I wanted to do this without needing to use `mut vec` and extending it. Greg clarified, though, that `[*include_globs, *(f"!{glob}" for glob in exclude_globs)]` is likely doing the same thing of creating a new list, only abstracted away from you. Rust is being more explicit here.